v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/transformer/GenericRecordTypeConvertor.java

```diff
@@ -0,0 +1,411 @@
+import org.apache.avro.Conversions;
+import org.apache.avro.LogicalTypes;
+import org.apache.avro.Schema;
+import org.apache.avro.data.TimeConversions;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.commons.codec.binary.Hex;
+import org.apache.commons.lang3.BooleanUtils;
+import org.apache.commons.lang3.math.NumberUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Convertor Class containing methods for type conversion of various AvroTypes to Spanner {@link
+ * Value} types.
+ */
+@AutoValue
+public abstract class GenericRecordTypeConvertor {
+  private static final Logger LOG = LoggerFactory.getLogger(GenericRecordTypeConvertor.class);
+
+  public abstract ISchemaMapper iSchemaMapper();
+
+  public abstract String namespace();
+
+  public static GenericRecordTypeConvertor create(ISchemaMapper iSchemaMapper, String namespace) {
+    return new AutoValue_GenericRecordTypeConvertor(iSchemaMapper, namespace);
+  }
+
+  /**
+   * This method takes in a generic record and returns a map between the Spanner column name and the
+   * corresponding Spanner column value. This handles the data conversion logic from a GenericRecord
+   * field to a spanner Value.
+   */
+  public Map<String, Value> transformChangeEvent(GenericRecord record, String srcTableName) {
+    Map<String, Value> result = new HashMap<>();
+    String spannerTableName = iSchemaMapper().getSpannerTableName(namespace(), srcTableName);
+    List<String> spannerColNames = iSchemaMapper().getSpannerColumns(namespace(), spannerTableName);
+    for (String spannerColName : spannerColNames) {
+      /**
+       * TODO: Handle columns that will not exist at source - synth id - shard id - multi-column
+       * transformations - auto-gen keys - Default columns - generated columns
+       */
+      String srcColName =
+          iSchemaMapper().getSourceColumnName(namespace(), spannerTableName, spannerColName);
+      Type spannerColumnType =
+          iSchemaMapper().getSpannerColumnType(namespace(), spannerTableName, spannerColName);
+      Value value = getSpannerValue(record, srcColName, spannerColumnType);
+      result.put(spannerColName, value);
+    }
+    return result;
+  }
+
+  /** Extract the field value from Generic Record and try to convert it to @spannerType. */
+  static Value getSpannerValue(GenericRecord record, String recordColName, Type spannerType) {
+    Object recordValue = record.get(recordColName);
+    Schema fieldSchema = record.getSchema().getField(recordColName).schema();
+    if (fieldSchema.getLogicalType() != null) {
+      // Logical should be converted to string.
+      recordValue = handleLogicalFieldType(recordColName, record);
+    } else if (fieldSchema.getType().equals(Schema.Type.RECORD)) {
+      // Record types should be converted to string.
+      recordValue = handleRecordFieldType(recordColName, record);
+    }
+    Schema.Type fieldType = fieldSchema.getType();
+    switch (spannerType.getCode()) {
```

Comment:
Can you give an example of what you mean?

Reply:
Maybe something like this - 

This will be in the constructor - 

```
SpannerDialect dialect = DialectRegistry.getDialectForRecord(record);
this.genericRecordConverter = recordConverterFactory.get(dialect);
```

Then later on - 

```
genericRecorderConverter.transformChangeEvent(record)
```

`genericRecorderConverter` will get an instance of either `GoogleSQLRecordCoverter` or `PGSQLRecordConverter`. They can internally utilize the same methods like `avroFieldToBoolean` for the actual underlying conversion to enable code re-use.

Reply:
Agreed.

Just had a syncup with @Deep1998 

A general recommendation internally is to use a static final ImmutableMap instead of switchcase for better readiblity.
This would be a map<Dialect, map<Type, ValueProviderInterface>>
The ValueProviderInterface implementation can be simplified by taking in labdas at build time if needed be.
I am doing something quite similar in my next PR which I shared.

Reply:
Love it!

<img src="https://media4.giphy.com/media/XreQmk7ETCak0/giphy.gif"/>